fathom/src/surface.rs

```diff
@@ -42,6 +47,383 @@ impl<'arena> Module<'arena, ByteRange> {
+                    let names: Vec<_> = dependencies[&id]
+                        .iter()
+                        .map(|id| interner.resolve(*id).unwrap())
+                        .collect();
+                    let this = interner.resolve(id).unwrap();
+                    dbg!((this, names));
+                    visit_item(
+                        id,
+                        top_level_scope,
+                        &dependencies,
+                        &mut visited,
+                        &mut output,
+                        &mut FxHashSet::default(),
+                    );
+                }
+                Item::ReportedError(_) => {}
+            }
+        }
+        output
+    }
+
+    fn collect_dependencies<'a>(&self, interner: &StringInterner, mut scope_chain: &mut ScopeChain<'a, 'arena>, dependencies: &mut DependencyMap) {
+        for item in self.items {
+            match item {
+                Item::Definition { label, type_, expr } => {
+                    // Add this item to scope
+                    scope_chain.define(label.1, item);
+
+                    // Collect deps
+                    let mut deps = Vec::new();
+                    if let Some(r#type) = type_ {
+                        preprocess_dependency(r#type, &mut scope_chain, &mut deps, interner);
+                    }
+                    scope_chain.push();
+                    preprocess_dependency(expr, &mut scope_chain, &mut deps, interner);
+                    scope_chain.pop();
+
+                    dependencies.insert(label.1, deps);
+                }
+                Item::ReportedError(_) => {}
+            }
+        }
+    }
+}
+
+fn visit_item<'a, 'arena>(
+    name: StringId,
+    top_level_scope: &ScopeFrame<'a, 'arena>,
+    dependencies: &DependencyMap,
+    visited: &mut FxHashSet<StringId>,
+    output: &mut Vec<&'a Item<'arena, ByteRange>>,
+    stack: &mut FxHashSet<StringId>,
+) {
+    if visited.contains(&name) {
+        return;
+    }
+
+    if stack.contains(&name) {
+        panic!("circular reference")
+    }
+
+    match top_level_scope.get(&name) {
+        Some(ScopeItem::Definition(item)) => {
+            stack.insert(name);
+            let deps = dependencies
+                .get(&name)
+                .map(|vec| vec.as_slice())
+                .unwrap_or_else(|| &[]);
+            for dep in deps {
+                visit_item(
+                    *dep,
+                    top_level_scope,
+                    dependencies,
+                    visited,
+                    output,
+                    stack,
+                );
+            }
+            stack.remove(&name);
+            visited.insert(name);
+            output.push(*item);
+        }
+        Some(ScopeItem::Name(name)) => {
+            visited.insert(*name);
+        }
+        Some(ScopeItem::Pending) | None => {}
+    }
+}
+
+#[derive(Copy, Clone)]
+enum ScopeItem<'a, 'arena> {
+    Definition(&'a Item<'arena, ByteRange>),
```

Comment:
I don't think this actually needs to carry the Item. I'll try removing it, which should tidy things up a bit.